ocaml/xapi/xapi_sr.ml

```diff
@@ -183,12 +183,44 @@ let probe ~__context ~host ~device_config ~_type ~sm_config =
 
   transform_storage_exn
     (fun () ->
-       match Client.SR.probe ~dbg ~queue ~device_config ~sm_config with
-       | Raw x -> x
-       | Probe _ as x -> Xmlrpc.to_string (rpc_of_probe_result x)
+       Client.SR.probe ~dbg ~queue ~device_config ~sm_config |> f
     )
 
-let probe_ext ~__context ~host ~device_config ~_type ~sm_config =
+let probe = call_probe ~f:(
+    function
+    | Storage_interface.Raw x -> x
+    | Storage_interface.Probe probe_results ->
+      (* Here we try to mimic the XML document structure returned in the Raw
+         case by SMAPIv1, for backwards-compatibility *)
+      let module T = Xmlm_tree in
+      let srs =
+        probe_results
+        (* We return the SRs found by probe *)
+        |> Xapi_stdext_std.Listext.List.filter_map
+          (fun r -> r.Storage_interface.sr)
+      in
+      let sr_info Storage_interface.{
+        name_label;
+        name_description;
+        total_space;
+        free_space;
+        clustered;
+        health} =
+        T.el
+          "SR"
+          [ T.el "size" [T.data @@ Int64.to_string total_space]
+          ; T.el "name_label" [T.data name_label]
+          ; T.el "name_description" [T.data name_description]
+          ]
+      in
+      let tree = T.el "SRlist" (List.map sr_info srs) in
+      let buf = Buffer.create 20 in
+      let output = Xmlm.make_output ~nl:true (`Buffer buf) in
+      T.output_doc ~tree ~output ~dtd:None;
```

Comment:
Ezxmlm would probably have some helpers for doing these 4 lines

Reply:
It does have some helpers for this